app-api/routes/dataset.py

```diff
@@ -1293,3 +1310,381 @@ async def update_metadata(
+            job
+            for job in load_jobs(session, dataset_id=id, user_id=user_id)
+            if job.extra_metadata.get("type") == JobType.POLICY_SYNTHESIS.value
+        ]
+
+        # If cluster_name is provided, filter jobs for that cluster
+        if cluster_name:
+            pending_jobs = [
+                job
+                for job in pending_jobs
+                if job.extra_metadata.get("cluster_name") == cluster_name
+            ]
+
+        # Try to cancel all pending jobs
+        for job in pending_jobs:
+            await cancel_job(session, job)
+
+        # Wait and check at most 2s
+        wait_and_check_timeout = 2
+        start_time = datetime.datetime.now()
+
+        # Wait and check for job status changes to be processed
+        while len(pending_jobs) > 0:
+            # Set off background task for checking jobs
+            asyncio.create_task(check_all_jobs())
+
+            # Get pending jobs again
+            jobs_query = [
+                job
+                for job in load_jobs(session, dataset_id=id, user_id=user_id)
+                if job.extra_metadata.get("type") == JobType.POLICY_SYNTHESIS.value
+            ]
+
+            # Apply cluster filter if specified
+            if cluster_name:
+                pending_jobs = [
+                    job
+                    for job in jobs_query
+                    if job.extra_metadata.get("cluster_name") == cluster_name
+                ]
+            else:
+                pending_jobs = jobs_query
+
+            # Wait a bit
+            await asyncio.sleep(0.5)
+
+            # Check if we waited too long overall
+            if (
+                datetime.datetime.now() - start_time
+            ).total_seconds() > wait_and_check_timeout:
+                break
+
+        # Return remaining jobs
+        return {
+            "jobs": [job.to_dict() for job in pending_jobs],
+        }
+
+
+@dataset.post("/jobs/cleanup")
+async def cleanup_jobs(
+    user_id: Annotated[UUID | None, Depends(UserIdentity)],
+    request: Request,
+    force: bool = False,
+    job_ids: List[str] = None,
+):
+    """
+    Administrative endpoint to clean up stale jobs.
+
+    Parameters:
+    - force: If true, clean up all jobs regardless of status
+    - job_ids: Optional list of specific job IDs to clean up
+    """
+    if job_ids:
+        # Clean up specific jobs by ID
+        with Session(db()) as session:
+            for job_id in job_ids:
+                jobs = session.query(DatasetJob).filter(DatasetJob.id == job_id).all()
+
+                for job in jobs:
+                    print(f"Manually cleaning up job {job.id}", flush=True)
+                    session.delete(job)
+
+            session.commit()
+        return {"message": f"Manually cleaned up {len(job_ids)} jobs"}
+
+    # Otherwise use the standard cleanup
+    await cleanup_stale_jobs(force_all=force)
+    return {"message": "Job cleanup complete. Check logs for details."}
```

Comment:
I think this endpoint may be insecure, as any user can now clean up any other users' jobs, right?